core/src/replay_stage.rs

```diff
@@ -783,37 +799,93 @@ impl ReplayStage {
+        // confirmed, what happens then? Should probably keep track of purged list and skip things
+        // in `duplicate_slots_to_repair` that have already been purged. Add test.
+        duplicate_slots_to_repair.retain(|(duplicate_slot, _correct_hash)| {
+            // Should not purge duplicate slots if there is currently a poh bank building
+            // on top of that slot, as BankingStage might still be referencing/touching that state
+            // concurrently.
+            // Luckily for us, because the fork choice rule removes duplicate slots from fork
+            // choice, and this function is called after:
+            // 1) We have picked a bank to reset to in `select_vote_and_reset_forks()`
+            // 2) And also called `reset_poh_recorder()`
+            // Then we should have reset to a fork that doesn't include the duplicate block,
+            // which means any working bank in PohRecorder that was built on that duplicate fork
+            // should have been cleared as well. However, if there is some violation of this guarantee,
+            // then log here
+            let is_poh_building_on_duplicate_fork = poh_bank_slot
+                .map(|poh_bank_slot| {
+                    ancestors
+                        .get(&poh_bank_slot)
+                        .expect("Poh bank should exist in BankForks and thus in ancestors map")
+                        .contains(&duplicate_slot)
+                })
+                .unwrap_or(false);
+
+            let did_purge_repair = {
+                if !is_poh_building_on_duplicate_fork {
+                    Self::purge_unconfirmed_duplicate_slot(
+                        *duplicate_slot,
+                        ancestors,
+                        descendants,
+                        progress,
+                        &root_bank,
+                        bank_forks,
+                        blockstore,
+                    );
+                    true
+                // TODO: Send signal to repair to repair the correct version of
+                // `duplicate_slot` with hash == `correct_hash`
+                } else {
+                    warn!(
+                        "PoH bank for slot {} is building on duplicate slot {}",
+                        poh_bank_slot.unwrap(),
+                        duplicate_slot
+                    );
+                    false
+                }
+            };
+
+            // If we purged/repaired, then no need to keep the slot in the set of pending work
+            !did_purge_repair
+        });
     }
 
-    #[allow(dead_code)]
     fn purge_unconfirmed_duplicate_slot(
         duplicate_slot: Slot,
         ancestors: &mut HashMap<Slot, HashSet<Slot>>,
         descendants: &mut HashMap<Slot, HashSet<Slot>>,
         progress: &mut ProgressMap,
+        root_bank: &Bank,
         bank_forks: &RwLock<BankForks>,
+        blockstore: &Blockstore,
```

Comment:
Should `duplicate_slots_tracker` be passed here too to have `duplicate_slot` removed from it?

Reply:
I don't think so, once a slot is put into `duplicate_slots_tracker`, it's to stay there forever until the root moves past that slot. The reason is if the goal of the tracker is to inform the `cluster_slot_state_verifier` state machine that this slot is duplicate and needs special handling. Just because we dumped one version of the slot, doesn't make this slot no longer a duplicate in the cluster.

Reply:
Makes sense, thanks